component/src/main/java/org/wso2/extension/siddhi/execution/streamingml/clustering/kmeans/KMeansIncrementalSPExtension.java

```diff
@@ -0,0 +1,305 @@
+import org.wso2.extension.siddhi.execution.streamingml.clustering.kmeans.util.DataPoint;
+import org.wso2.extension.siddhi.execution.streamingml.clustering.kmeans.util.KMeansModel;
+import org.wso2.extension.siddhi.execution.streamingml.clustering.kmeans.util.KMeansModelHolder;
+import org.wso2.siddhi.annotation.Example;
+import org.wso2.siddhi.annotation.Extension;
+import org.wso2.siddhi.annotation.Parameter;
+import org.wso2.siddhi.annotation.ReturnAttribute;
+import org.wso2.siddhi.annotation.util.DataType;
+import org.wso2.siddhi.core.config.SiddhiAppContext;
+import org.wso2.siddhi.core.event.ComplexEventChunk;
+import org.wso2.siddhi.core.event.stream.StreamEvent;
+import org.wso2.siddhi.core.event.stream.StreamEventCloner;
+import org.wso2.siddhi.core.event.stream.populater.ComplexEventPopulater;
+import org.wso2.siddhi.core.executor.ConstantExpressionExecutor;
+import org.wso2.siddhi.core.executor.ExpressionExecutor;
+import org.wso2.siddhi.core.executor.VariableExpressionExecutor;
+import org.wso2.siddhi.core.query.processor.Processor;
+import org.wso2.siddhi.core.query.processor.stream.StreamProcessor;
+import org.wso2.siddhi.core.util.config.ConfigReader;
+import org.wso2.siddhi.query.api.definition.AbstractDefinition;
+import org.wso2.siddhi.query.api.definition.Attribute;
+import org.wso2.siddhi.query.api.exception.SiddhiAppValidationException;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * performs kmeans with incremental update
+ */
+@Extension(
+        name = "kMeansIncremental",
+        namespace = "streamingml",
+        description = "Performs K-Means clustering on a streaming data set. Data points can be of any dimension and " +
+                "the dimensionality is calculated from number of parameters. " +
+                "All data points to be processed by an instance of class Clusterer should be of the same " +
+                "dimensionality. The Euclidean distance is taken as the distance metric. " +
+                "The algorithm resembles Sequential K-Means Clustering at " +
+                "https://www.cs.princeton.edu/courses/archive/fall08/cos436/Duda/C/sk_means.htm ",
+        parameters = {
+                @Parameter(
+                        name = "model.name",
+                        description = "The name for the model that is going to be created/reused for prediction",
+                        type = {DataType.STRING}
+                ),
+                @Parameter(
+                        name = "decay.rate",
+                        description = "this is the decay rate of old data compared to new data. " +
+                                "Value of this will be in [0,1]. 0 means only old data used and" +
+                                "1 will mean that only new data is used",
+                        optional = true,
+                        type = {DataType.FLOAT},
```

Comment:
Why is this float? I think for consistency, we should stick with double.

Reply:
done